File: python/modulus_adapter_langchain/modulus_adapter_langchain/handler.py
import time
import asyncio
from typing import Any, Optional
from uuid import UUID
from langchain_core.callbacks import BaseCallbackHandler
from modulus_sdk import ModulusClient


class ModulusCallbackHandler(BaseCallbackHandler):
    def __init__(self, api_key: str, agent_name: str, endpoint: str = "https://ingest.modulus.dev",
                 commit_sha: Optional[str] = None, known_tools: Optional[list[str]] = None):
        super().__init__()
        self.client = ModulusClient(api_key=api_key, agent_name=agent_name, endpoint=endpoint,
                                    framework="langchain", commit_sha=commit_sha, known_tools=known_tools)
        self._span_by_run_id: dict[str, str] = {}
        self._tool_runs: dict[str, tuple[str, Any, int]] = {}
        self._llm_runs: dict[str, tuple[str, int]] = {}
        self._flush_handle: Optional[asyncio.TimerHandle] = None

    def _parent_span_id(self, parent_run_id: Optional[UUID]) -> Optional[str]:
        return self._span_by_run_id.get(str(parent_run_id)) if parent_run_id else None

    def _schedule_flush(self):
        #idle flush. (`await handler.client.flush()` explicitly)
        try:
            loop = asyncio.get_running_loop()
        except RuntimeError:
            return
        if self._flush_handle:
            self._flush_handle.cancel()
        self._flush_handle = loop.call_later(
            5, lambda: asyncio.ensure_future(self.client.flush()))

    def on_chain_start(self, serialized: dict, inputs: dict, *, run_id: UUID, parent_run_id: Optional[UUID] = None, **kwargs: Any) -> None:
        span_id = self.client.start_span("chain", int(
            time.time() * 1000), self._parent_span_id(parent_run_id))
        self._span_by_run_id[str(run_id)] = span_id

    def on_chain_end(self, outputs: dict, *, run_id: UUID, **kwargs: Any) -> None:
        span_id = self._span_by_run_id.get(str(run_id))
        if span_id:
            self.client.end_span(span_id, "ok")
        self._schedule_flush()

    def on_chain_error(self, error: BaseException, *, run_id: UUID, **kwargs: Any) -> None:
        span_id = self._span_by_run_id.get(str(run_id))
        if span_id:
            self.client.end_span(span_id, "error")
        self._schedule_flush()

    def on_tool_start(self, serialized: dict, input_str: str, *, run_id: UUID, **kwargs: Any) -> None:
        self._tool_runs[str(run_id)] = (serialized.get(
            "name", "unknown_tool"), input_str, int(time.time() * 1000))

    def on_tool_end(self, output: Any, *, run_id: UUID, parent_run_id: Optional[UUID] = None, **kwargs: Any) -> None:
        tool_name, tool_input, started_at = self._tool_runs.pop(
            str(run_id), ("unknown_tool", None, int(time.time() * 1000)))
        self.client.record_tool_call(
            tool_name, tool_input, "success", output, started_at, self._parent_span_id(parent_run_id))

    def on_tool_error(self, error: BaseException, *, run_id: UUID, parent_run_id: Optional[UUID] = None, **kwargs: Any) -> None:
        tool_name, tool_input, started_at = self._tool_runs.pop(
            str(run_id), ("unknown_tool", None, int(time.time() * 1000)))
        self.client.record_tool_call(tool_name, tool_input, "error", {"message": str(
            error)}, started_at, self._parent_span_id(parent_run_id))

    def on_llm_start(self, serialized: dict, prompts: list[str], *, run_id: UUID, **kwargs: Any) -> None:
        model_name = serialized.get("kwargs", {}).get(
            "model", serialized.get("name", "unknown_model"))
        self._llm_runs[str(run_id)] = (model_name, int(time.time() * 1000))

    def on_llm_end(self, response: Any, *, run_id: UUID, parent_run_id: Optional[UUID] = None, **kwargs: Any) -> None:
        model_name, started_at = self._llm_runs.pop(
            str(run_id), ("unknown_model", int(time.time() * 1000)))
        try:
            text = response.generations[0][0].text
        except (AttributeError, IndexError):
            text = ""
        self.client.record_model_call(model_name, text, "success", started_at, parent_span_id=self._parent_span_id(parent_run_id))

    def on_llm_error(self, error: BaseException, *, run_id: UUID, parent_run_id: Optional[UUID] = None, **kwargs: Any) -> None:
        model_name, started_at = self._llm_runs.pop(str(run_id), ("unknown_model", int(time.time() * 1000)))
        self.client.record_model_call(model_name, str(error), "error", started_at, parent_span_id=self._parent_span_id(parent_run_id))

    def on_retriever_start(self, serialized: dict, query: str, *, run_id: UUID, **kwargs: Any) -> None:
        self._tool_runs[f"retriever:{run_id}"] = (
            "__retriever__", query, int(time.time() * 1000))

    def on_retriever_end(self, documents: list, *, run_id: UUID, parent_run_id: Optional[UUID] = None, **kwargs: Any) -> None:
        _, query, started_at = self._tool_runs.pop(f"retriever:{run_id}", ("__retriever__", "", int(time.time() * 1000)))
        scores = [d.metadata.get("score") for d in documents if getattr(d, "metadata", None) and isinstance(d.metadata.get("score"), (int, float))]
        span_id = self.client.start_span("retrieval", started_at, self._parent_span_id(parent_run_id))

        self.client.tag_span(span_id, "modulus.retrieval.query", query)
        self.client.tag_span(
            span_id, "modulus.retrieval.result_count", len(documents))
        if scores:
            self.client.tag_span(span_id, "modulus.retrieval.top_score", max(scores))
        self.client.end_span(span_id, "ok")

File: packages/adapter-langchain/src/handler.ts
import { BaseCallbackHandler } from "@langchain/core/callbacks/base";
import type { Serialized } from "@langchain/core/load/serializable";
import { ModulusClient } from "@modulus/sdk";
import type { LLMResult } from "@langchain/core/outputs";
// import type { AgentAction, AgentFinish } from "@langchain/core/agents";
// import { ChainValues } from "@langchain/core/utils/types";

interface ModulusHandlerConfig {
	apiKey: string;
	agentName: string;
	endpoint?: string;
	commitSha?: string;
	knownTools?: string[];
}

interface ToolRunState {
	toolName: string;
	input: string;
	startedAt: number;
}

const IDLE_FLUSH_MS = 5_000;

export class ModulusCallbackHandler extends BaseCallbackHandler {
	name = "modulus_callback_handler";
	protected client: ModulusClient;
	protected spanIdByRunId = new Map<string, string>();
	private toolRuns = new Map<string, ToolRunState>();
	private llmRuns = new Map<string, { modelName: string; startedAt: number }>();
	private flushTimer: ReturnType<typeof setTimeout> | null = null;
	private retrieverRuns = new Map<
		string,
		{ query: string; startedAt: number }
	>();

	// constructor
	constructor(config: ModulusHandlerConfig) {
		super();
		this.client = new ModulusClient({
			apiKey: config.apiKey,
			agentName: config.agentName,
			endpoint: config.endpoint,
			commitSha: config.commitSha,
			knownTools: config.knownTools,
		});
	}

	// auto log (failure handling)
	private scheduleFlush() {
		if (this.flushTimer) clearTimeout(this.flushTimer);

		this.flushTimer = setTimeout(() => {
			void this.client.flush().catch(() => {});
			this.flushTimer = null;
		}, IDLE_FLUSH_MS);

		this.flushTimer.unref?.();
	}

	private parentSpanId(parentRunId?: string) {
		return parentRunId ? this.spanIdByRunId.get(parentRunId) : undefined;
	}

	// langchain chain start
	async handleChainStart(
		_chain: Serialized,
		_inputs: unknown,
		runId: string,
		parentRunId?: string,
	) {
		this.spanIdByRunId.set(
			runId,
			this.client.startSpan("chain", {
				startedAt: Date.now(),
				parentSpanId: this.parentSpanId(parentRunId),
			}),
		);
	}

	// handle chain end
	async handleChainEnd(_outputs: unknown, runId: string) {
		const spanId = this.spanIdByRunId.get(runId);
		if (spanId) this.client.endSpan(spanId, { status: "ok" });

		this.scheduleFlush();
	}

	// chain error handling
	async handleChainError(_err: Error, runId: string) {
		const spanId = this.spanIdByRunId.get(runId);
		if (spanId) this.client.endSpan(spanId, { status: "error" });

		this.scheduleFlush();
	}

	// retriever call start
	async handleRetrieverStart(
		_retriever: Serialized,
		query: string,
		runId: string,
	) {
		this.retrieverRuns.set(runId, { query, startedAt: Date.now() });
	}

	// retriever call end
	async handleRetrieverEnd(
		documents: Array<{ metadata?: Record<string, unknown> }>,
		runId: string,
		parentRunId?: string,
	) {
		const run = this.retrieverRuns.get(runId);
		const topScore = documents
			.map((d) =>
				typeof d.metadata?.score === "number" ? d.metadata.score : undefined,
			)
			.filter((s): s is number => s !== undefined)
			.sort((a, b) => b - a)[0];

		this.client.recordRetrieval(run?.query ?? "", {
			topScore,
			resultCount: documents.length,
			startedAt: run?.startedAt ?? Date.now(),
			parentSpanId: this.parentSpanId(parentRunId),
		});
		this.retrieverRuns.delete(runId);
	}

	// tool call start
	async handleToolStart(tool: Serialized, input: string, runId: string) {
		this.toolRuns.set(runId, {
			toolName: String(tool.id.at(-1) ?? "unknown_tool"),
			input,
			startedAt: Date.now(),
		});
	}

	// tool call end
	async handleToolEnd(output: string, runId: string, parentRunId?: string) {
		const run = this.toolRuns.get(runId);
		this.client.recordToolCall(run?.toolName ?? "unknown_tool", run?.input, {
			status: "success",
			response: output,
			startedAt: run?.startedAt ?? Date.now(),
			parentSpanId: this.parentSpanId(parentRunId),
		});
		this.toolRuns.delete(runId);
	}

	// tool call error handling
	async handleToolError(err: Error, runId: string, parentRunId?: string) {
		const run = this.toolRuns.get(runId);
		this.client.recordToolCall(run?.toolName ?? "unknown_tool", run?.input, {
			status: "error",
			response: { message: err.message },
			startedAt: run?.startedAt ?? Date.now(),
			parentSpanId: this.parentSpanId(parentRunId),
		});
		this.toolRuns.delete(runId);
	}

	// model(llm) call start
	async handleLLMStart(llm: Serialized, _prompts: string[], runId: string) {
		this.llmRuns.set(runId, {
			modelName: String(llm.id.at(-1) ?? "unknown_model"),
			startedAt: Date.now(),
		});
	}

	// model(llm) call end
	async handleLLMEnd(output: LLMResult, runId: string, parentRunId?: string) {
		const run = this.llmRuns.get(runId);
		const text = output?.generations?.[0]?.[0]?.text ?? "";
		this.client.recordModelCall(run?.modelName ?? "unknown_model", {
			output: text,
			status: "success",
			startedAt: run?.startedAt ?? Date.now(),
			parentSpanId: this.parentSpanId(parentRunId),
		});
		this.llmRuns.delete(runId);
	}

	// model(llm) error handling
	async handleLLMError(err: Error, runId: string, parentRunId?: string) {
		const run = this.llmRuns.get(runId);
		this.client.recordModelCall(run?.modelName ?? "unknown_model", {
			output: err.message,
			status: "error",
			startedAt: run?.startedAt ?? Date.now(),
			parentSpanId: this.parentSpanId(parentRunId),
		});
		this.llmRuns.delete(runId);
	}
}